cirq-core/cirq/ops/kraus_channel.py

```diff
@@ -0,0 +1,102 @@
+from typing import Any, Dict, Iterable, Tuple, Union
+import numpy as np
+
+from cirq import linalg, protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+class KrausChannel(raw_types.Gate):
+    """A generic channel that can record the index of its selected operator.
+
+    Args:
+        kraus_ops: a list of Kraus operators, formatted as numpy array.
+        key: an optional measurement key string for this channel. Simulations
+            which select a single Kraus operator to apply will store the index
+            of that operator in the measurement result list with this key.
+        validate: if True, validate that `kraus_ops` describe a valid channel.
+            This validation can be slow; prefer pre-validating if possible.
+    """
+
+    def __init__(
+        self,
+        kraus_ops: Iterable[np.ndarray],
+        key: Union[str, value.MeasurementKey, None] = None,
+        validate: bool = False,
+    ):
+        kraus_ops = list(kraus_ops)
+        if not kraus_ops:
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
```

Comment:
Many interesting channels aren't n-qubit channels. Two interesting examples that come to mind: Werner-Holevo channel for odd dimension and channels that model leakage. It would be nice to support these.

Reply:
AFAIK Cirq doesn't have support for qubits which change dimension over the course of a circuit. While I agree supporting these could be worthwhile, I think it's out of scope for this PR. If this is something you'd like to see, could you open an issue to track it?

Reply:
(I think you meant to write this in response to the other comment?)

I agree with you: it's not even clear right now whether channels with different input and output spaces should be represented as acting on one qudit and changing its dimension or discarding the input qudit and targeting a new qudit. Unfortunately, this may be hard to shoehorn into cirq's model. Definitely out of scope here.

That said, let's make it clear in the comments that this is out assumption and that as a consequence all Kraus operators are square matrices.

---

The comment about supporting non-qubit channels remains, so not marking this as resolved (but will mark the other one as resolved).

Reply:
Sorry, I missed that these were distinct issues. The latest change includes a note about both (i.e. KrausChannel only supports qubit-based square operators).